### DISCUSSION TITLE
How to take extension arguments?

### DISCUSSION THREAD

**Discussion by steverice** [mirumee/ariadne#873] [Q&A]
A bit confused on how an Ariadne extension is supposed to receive arguments upon instantiation.

There's an example of it in https://github.com/mirumee/ariadne/pull/608 with the `trace_default_resolver` argument, shown in [the docs](https://github.com/mirumee/ariadne-website/pull/94/files) as being called like:
```python
extensions=[ApolloTracingExtension(trace_default_resolver=True)]
```

When I try this however, I'm getting an error:
```
  File "/usr/local/lib/python3.7/dist-packages/ariadne/extensions.py", line 21, in __init__
    self.extensions = tuple(ext() for ext in extensions)
  File "/usr/local/lib/python3.7/dist-packages/ariadne/extensions.py", line 21, in <genexpr>
    self.extensions = tuple(ext() for ext in extensions)
TypeError: 'ApolloTracingExtension' object is not callable
```

Indeed, this error makes sense — although the `ExtensionManager` [purports to take an `ExtensionList`](https://github.com/mirumee/ariadne/blob/9f410261080933883e91ae887499c2f2e343a476/ariadne/extensions.py#L15) (which can [be an instance of an `Extension`](https://github.com/mirumee/ariadne/blob/9f410261080933883e91ae887499c2f2e343a476/ariadne/types.py#L46) as well as a class), it then [immediately tries to call each extension](https://github.com/mirumee/ariadne/blob/master/ariadne/extensions.py#L21). It also does not seem that `Extension` instances are expected to be callable.

It seems like I'm missing something, rather than this being a bug; especially as the [code that calls extensions](https://github.com/mirumee/ariadne/commit/ef6d381cdbcfcdbdf3944f4e8edec0ff031b6804) predates the implementation of that extension with an argument. So how is this supposed to work?

**Comment by rafalp** [comment]
You are not wrong, docs are incorrect here, because extension class is instantiated per query to let it store state. We should introduce utility functions for those like we do for [query cost validator](https://github.com/mirumee/ariadne/blob/master/ariadne/validation/query_cost.py#L341), that reduce partials keeping configuration from dev

**Reply by steverice** [reply]
Thanks @rafalp ! So would the suggestion for now be to use a partial?
I'm noticing now that I read the type of `ExtensionList` incorrectly — it indeed suggests that the extension must either be a class or a partial.
